### DISCUSSION TITLE
Please help with multivariate time series analysis in gluonts

### DISCUSSION THREAD

**Discussion by pratikgehlott** [awslabs/gluonts#1428] [General]
I need urgent help in this part, please can anyone help?

**Comment by mbohlkeschneider** [comment]
Hi @Pratik325, 

can you be more precise? What exactly is the issue? 

Best regards.

**Reply by pratikgehlott** [reply]
yes so @mbohlkeschneider , we previously also had talk on this issue https://github.com/awslabs/gluon-ts/issues/652 .

So my issue is i am not able to bring my data in the format that gluonts requires.

**Reply by pratikgehlott** [reply]
so @mbohlkeschneider consider for example we have 
columns = [date , feature1 , feature2 , feature3 , target ],
now i need to predict target.

so according to you how we will bring this data in the format that gluonts requires?

**Reply by mbohlkeschneider** [reply]
I see. It depends what model you want to use and what your setup is specifically. 

If you want to forecast the target and you know the feature values in the future, I would suggest to encode the features as `feat_dynamic_real` and use DeepAR. 

If you don't know the feature values in the future, you can try MQ-CNN which only accepts past values. Alternatively, you could try using a multivariate forecasting model, but if we have only looked in the papers into the homogeneous case (all dimensions of the multivariate time series are the same type, say sales). 

So what it is what you want to do exactly?

**Reply by pratikgehlott** [reply]
The first case, I know all the features.

**Reply by pratikgehlott** [reply]
**I would suggest to encode the features as feat_dynamic_real and use DeepAR.**


This is part i am trying to understand from past 48 hours.

**Reply by pratikgehlott** [reply]
can u please elaborate on this only, how to do it.

**Reply by pratikgehlott** [reply]
why @mbohlkeschneider you always leave a convo in between, this is 3rd time with you. :(

**Reply by mbohlkeschneider** [reply]
Hi @Pratik325, 

Here is a snipped on how to setup a dummy dataset with dynamic features. Basically, you have to convert your data into this format. Let me know if there are more questions. For more information, please refer to the extended tutorial: https://ts.gluon.ai/tutorials/forecasting/extended_tutorial.html

```
import pprint

from gluonts.dataset.repository.datasets import get_dataset, dataset_recipes
from gluonts.evaluation import Evaluator
from gluonts.evaluation.backtest import make_evaluation_predictions
from gluonts.model.deepar import DeepAREstimator
from gluonts.mx.trainer import Trainer
from gluonts.dataset.common import ListDataset
import numpy as np

data_entry_train = {"target": np.linspace(0,1,100), "feat_dynamic_real": [np.linspace(0,1,100)], "start":"2020-01-01"}

dataset_train = ListDataset([data_entry_train]*100, freq='H')

data_entry_test = {"target": np.linspace(0,1,124), "feat_dynamic_real": [np.linspace(0,1,124)], "start":"2020-01-01"}

dataset_test = ListDataset([data_entry_test]*100, freq='H')


estimator = DeepAREstimator(
    prediction_length=24,
    freq="H",
    trainer=Trainer(epochs=5, num_batches_per_epoch=10),
    use_feat_dynamic_real=True
)

predictor = estimator.train(dataset_train)

forecast_it, ts_it = make_evaluation_predictions(
    dataset_test, predictor=predictor, num_samples=100
)

agg_metrics, item_metrics = Evaluator()(
    ts_it, forecast_it, num_series=len(dataset_test)
)

pprint.pprint(agg_metrics)
```

**Reply by MAYUKH97** [reply]
Hi @mbohlkeschneider I have a follow up question So, I have a data in the form of hierarchy  and want to forecast for the granular level. For example data = ["food" ,"food_category", "revenue"], So in food we have different types of items and we segregate on basis of chinese, thai, india and then in chinese we may have like noodles, pasta, etc. so as every segment is connected so how do we forecast revenue the granular level for each individual items.

**Reply by mbohlkeschneider** [reply]
Hi @MAYUKH97, 

you can encode them as static cat features using standard label encoding. We currently don't have a "real" hierarchical forecasting algorithm in GluonTS currently.

**Reply by pratikgehlott** [reply]
hey @mbohlkeschneider can we handle multiple row having same dates in gluonts?

**Reply by mbohlkeschneider** [reply]
Hi @Pratik325, 

as mentioned above, this depends on your forecasting scenario and how your input data is stored. Note that GluonTS does not provide any tooling to convert data into GluonTS format - this has to be done by the user. If you have specific questions about the GluonTS data format, I'll be happy to answer them.
